cpp/examples/tx_send.cpp

```diff
@@ -0,0 +1,175 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#include "options.hpp"
+
+#include <proton/connection.hpp>
+#include <proton/container.hpp>
+#include <proton/message.hpp>
+#include <proton/message_id.hpp>
+#include <proton/messaging_handler.hpp>
+#include <proton/types.hpp>
+#include <proton/transaction.hpp>
+
+#include <iostream>
+#include <map>
+#include <string>
+
+#include <chrono>
+#include <thread>
+
+class tx_send : public proton::messaging_handler, proton::transaction_handler {
+  private:
+    proton::sender sender; 
+    std::string url;
+    int total;
+    int batch_size;
+    int sent;
+    int batch_index = 0;
+    int current_batch = 0;
+    int committed = 0;
+    int confirmed = 0;
+
+    proton::session session;
+    proton::transaction transaction;
+  public:
+    tx_send(const std::string &s, int c, int b):
+        url(s), total(c), batch_size(b), sent(0) {}
+
+    void on_container_start(proton::container &c) override {
+        sender = c.open_sender(url);
+    }
+
+    void on_session_open(proton::session &s) override {
+        session = s;
+        std::cout << "    [on_session_open] declare_txn started..." << std::endl;
+        s.declare_transaction(*this);
+        std::cout << "    [on_session_open] declare_txn ended..." << std::endl;
+    }
+
+    void on_transaction_declare_failed(proton::transaction) {}
+    void on_transaction_commit_failed(proton::transaction t) {
+        std::cout << "Transaction Commit Failed" << std::endl;
+        t.connection().close();
+        exit(-1);
+    }
+
+    void on_transaction_declared(proton::transaction t) override {
+        std::cout << "[on_transaction_declared] txn called " << (&t)
+                  << std::endl;
+        std::cout << "[on_transaction_declared] txn is_empty " << (t.is_empty())
+                  << "\t" << transaction.is_empty() << std::endl;
+        transaction = t;
+
+        send(sender);
+    }
+
+    void on_sendable(proton::sender &s) override {
+        std::cout << "    [OnSendable] transaction: " << &transaction
+                  << std::endl;
+        send(s);
+    }
+
+    void send(proton::sender &s) {
+        static int unique_id = 10000;
+        while (!transaction.is_empty() && sender.credit() &&
+               (committed + current_batch) < total) {
+            proton::message msg;
+            std::map<std::string, int> m;
+            m["sequence"] = committed + current_batch;
+
+            msg.id(unique_id++);
+            msg.body(m);
+            std::cout << "##### [example] transaction send msg: " << msg
+                      << std::endl;
+            transaction.send(sender, msg);
+            current_batch += 1;
+            if(current_batch == batch_size)
+            {
+                std::cout << " >> Txn attempt commit" << std::endl;
+                if (batch_index % 2 == 0) {
```

Comment:
I see that every second batch is aborted. I'm not sure if that's intended, though it works as expected (python example is not doing any aborts).